go-controller/pkg/node/gateway_localnet.go

```diff
@@ -239,89 +246,374 @@ func localnetIPTablesHelper(subnet *net.IPNet) (util.IPTablesHelper, error) {
+				"-d", externalIP,
+				"--dport", fmt.Sprintf("%v", svcPort.Port),
+				"-j", "ACCEPT",
+			},
+		},
+	}
+}
+
+func (npw *localnetNodePortWatcherData) networkHasAddress(ipNet net.IP) bool {
+	for _, net := range npw.localAddrSet {
+		if net.Contains(ipNet) {
+			return true
+		}
+	}
+	return false
+}
+
+func (npw *localnetNodePortWatcherData) emitPortClaimEvent(svc *kapi.Service, port int32, err error) {
+	serviceRef := kapi.ObjectReference{
+		Kind:      "Service",
+		Namespace: svc.Namespace,
+		Name:      svc.Name,
+	}
+	npw.recorder.Eventf(&serviceRef, kapi.EventTypeWarning, "PortClaim", "Service: %s specifies node local port: %v, but port cannot be bound to: %v", svc.Name, port, err)
+	klog.Warningf("PortClaim for svc: %s on port: %v, err: %v", svc.Name, port, err)
 }
 
 func (npw *localnetNodePortWatcherData) addService(svc *kapi.Service) error {
-	if !util.ServiceTypeHasNodePort(svc) {
-		return nil
+	iptRules := []iptRule{}
+	for _, port := range svc.Spec.Ports {
+		if util.ServiceTypeHasNodePort(svc) {
+			if err := util.ValidatePort(port.Protocol, port.NodePort); err != nil {
+				klog.Warningf("invalid service node port %s, err: %v", port.Name, err)
+				continue
+			}
+			if err := npw.openLocalPort(port.NodePort, port.Protocol, svc); err != nil {
+				continue
+			}
+			iptRules = append(iptRules, npw.nodePortIPTRules(port, npw.gatewayIP)...)
+			klog.V(5).Infof("will add iptables rule for NodePort: %v and protocol: %v", port.NodePort, port.Protocol)
+		}
+		for _, externalIP := range svc.Spec.ExternalIPs {
+			if err := util.ValidatePort(port.Protocol, port.Port); err != nil {
+				klog.Warningf("invalid service port %s, err: %v", port.Name, err)
+				continue
+			}
+			if _, exists := npw.localAddrSet[externalIP]; exists {
+				if !util.IsClusterIPSet(svc) {
+					serviceRef := kapi.ObjectReference{
+						Kind:      "Service",
+						Namespace: svc.Namespace,
+						Name:      svc.Name,
+					}
+					npw.recorder.Eventf(&serviceRef, kapi.EventTypeWarning, "UnsupportedServiceDefinition", "Unsupported service definition, headless service: %s with a local ExternalIP is not supported by ovn-kubernetes in local gateway mode", svc.Name)
+					klog.Warningf("UnsupportedServiceDefinition event for service %s in namespace %s", svc.Name, svc.Namespace)
+					continue
+				}
+				if err := npw.openLocalPort(port.Port, port.Protocol, svc); err != nil {
+					continue
+				}
+				iptRules = append(iptRules, npw.externalIPTRules(port, externalIP, svc.Spec.ClusterIP)...)
+				klog.V(5).Infof("will add iptables rule for ExternalIP: %s", externalIP)
+				continue
+			}
+			if npw.networkHasAddress(net.ParseIP(externalIP)) {
+				klog.V(5).Infof("ExternalIP: %s is reachable through one of the interfaces on this node, will skip setup", externalIP)
+				continue
+			}
+			if stdout, stderr, err := util.RunIP("route", "add", externalIP, "via", npw.gatewayIP, "dev", localnetGatewayNextHopPort, "table", localnetGatwayExternalIDTable); err != nil {
```

Comment:
this should be `ip route replace`, which is idempotent.

Reply:
Ok, changed it